ui: Allow callbook lookups for WSJT-X contacts even without grid square

sjwoodr opened the pull request:
This change is to allow WSJT-X format broadcast packets to trigger a callbook lookup even when the grid square is not specified. This is useful for non FT8 content. For example, realtime contest logs being sent from WriteLog to Qlog via the WSJT-X protocol.  It is safe to call finalizeCallsignEdit() even without the grid square because the CallbookManager will cache lookups and prevent redundant lookups when WSJT-X sends multiple status type messages.

foldynl commented:
Unfortunately, this change breaks the essence of why it was there. And that is the fact that if you have a QSO from WSJT and it doesn't have a Grid, you can't tell whether the received Callbook entry corresponds to the current position of the DX Station. That's why the condition "to have a grid" was there.

sjwoodr commented:
> Unfortunately, this change breaks the essence of why it was there. And that is the fact that if you have a QSO from WSJT and it doesn't have a Grid, you can't tell whether the received Callbook entry corresponds to the current position of the DX Station. That's why the condition "to have a grid" was there.

OK, that makes sense.  Right now I have a work-around, and that is sending grid AA00 if it is not filled in.  That's in the middle of the ocean and unlikely to be valid.  This will trigger a callbook lookup in Qlog ... but it results in the AA00 being logged (the callbook lookup won't override it).  

What would you say about allowing the callbook to override the grid if it was something like AA00 or even something invalid like `0000` ?

sjwoodr commented:
@foldynl How about a change like this instead?  
```
diff --git a/ui/NewContactWidget.cpp b/ui/NewContactWidget.cpp
index 9d5418c0..1926cb3c 100644
--- a/ui/NewContactWidget.cpp
+++ b/ui/NewContactWidget.cpp
@@ -2939,7 +2939,17 @@ void NewContactWidget::prepareWSJTXQSO(const QString &receivedCallsign,

     callsign = receivedCallsign;
     ui->callsignEdit->setText(receivedCallsign);
-    uiDynamic->gridEdit->setText(grid);
+
+    // WSJT-X packets with "XXXX" as a grid:
+    // We should trigger callbook lookup but not persist XXXX in the UI
+    bool shouldTriggerCallbookLookup = (grid == "XXXX");
+    QString gridToUse = shouldTriggerCallbookLookup ? QString() : grid;
+
+    if ( !gridToUse.isEmpty() )
+    {
+        uiDynamic->gridEdit->setText(gridToUse);
+    }
+
     checkDupe();
     setDxccInfo(receivedCallsign);
     queryPota();
@@ -2951,9 +2961,10 @@ void NewContactWidget::prepareWSJTXQSO(const QString &receivedCallsign,
     // 1) prev Callsign empty grid
     // 2) new Callsign empty grid
     // 3) new Calllsign, new gris
-    if ( !grid.isEmpty() )
+    // If grid is "XXXX", treat it as a trigger to perform callbook lookup
+    if ( !gridToUse.isEmpty() || shouldTriggerCallbookLookup )
     {
-        useFieldsFromPrevQSO(callsign, grid);
+        useFieldsFromPrevQSO(callsign, gridToUse);
         finalizeCallsignEdit();
     }
 }
```

sjwoodr commented:
@foldynl - OK let's look at what your suggestion would look like:

Key Changes:

1. ui/WsjtxWidget.h & ui/WsjtxWidget.cpp: Updated the callsignSelected signal to include the application ID field
    - Signal now: void callsignSelected(QString callsign, QString grid, QString id)
2. ui/NewContactWidget.h & ui/NewContactWidget.cpp: Updated prepareWSJTXQSO to accept and use the ID parameter
    - Method signature: void prepareWSJTXQSO(const QString &receivedCallsign, const QString &grid, const  QString &id)
    - Added WRITELOG-specific logic: If id.contains("WRITELOG"), always trigger callbook lookup
    - Grid handling: Only sets grid in UI if not empty

How it works:
  - When a packet comes from an application with ID containing "WRITELOG", QLog will always trigger the callbook lookup (via useFieldsFromPrevQSO and finalizeCallsignEdit)
  - The grid field is handled normally - if it's provided and not empty, it's displayed in the UI
  - This follows the same pattern used for JTDX timezone handling (lines 299 and 342 in core/WsjtxUDPReceiver.cpp:299 and core/WsjtxUDPReceiver.cpp:342)
 
```
diff --git a/ui/NewContactWidget.cpp b/ui/NewContactWidget.cpp
index 9d5418c0..7f39acf9 100644
--- a/ui/NewContactWidget.cpp
+++ b/ui/NewContactWidget.cpp
@@ -2915,11 +2915,12 @@ void NewContactWidget::fillCallsignGrid(const QString &callsign, const QString &
 }

 void NewContactWidget::prepareWSJTXQSO(const QString &receivedCallsign,
-                                       const QString &grid)
+                                       const QString &grid,
+                                       const QString &id)
 {
     FCT_IDENTIFICATION;

-    qCDebug(function_parameters) << receivedCallsign << grid;
+    qCDebug(function_parameters) << receivedCallsign << grid << id;

     if ( isManualEnterMode )
     {
@@ -2939,7 +2940,12 @@ void NewContactWidget::prepareWSJTXQSO(const QString &receivedCallsign,

     callsign = receivedCallsign;
     ui->callsignEdit->setText(receivedCallsign);
-    uiDynamic->gridEdit->setText(grid);
+
+    if ( !grid.isEmpty() )
+    {
+        uiDynamic->gridEdit->setText(grid);
+    }
+
     checkDupe();
     setDxccInfo(receivedCallsign);
     queryPota();
@@ -2951,7 +2957,9 @@ void NewContactWidget::prepareWSJTXQSO(const QString &receivedCallsign,
     // 1) prev Callsign empty grid
     // 2) new Callsign empty grid
     // 3) new Calllsign, new gris
-    if ( !grid.isEmpty() )
+    // WRITELOG workaround: Similar to JTDX, check application ID
+    // If ID contains "WRITELOG", always trigger callbook lookup
+    if ( !grid.isEmpty() || id.contains("WRITELOG") )
     {
         useFieldsFromPrevQSO(callsign, grid);
         finalizeCallsignEdit();
diff --git a/ui/NewContactWidget.h b/ui/NewContactWidget.h
index e79b4a5a..1ec99c1f 100644
--- a/ui/NewContactWidget.h
+++ b/ui/NewContactWidget.h
@@ -236,7 +236,7 @@ public slots:
     void readGlobalSettings();
     void tuneDx(const DxSpot &spot);
     void fillCallsignGrid(const QString &callsign, const QString& grid);
-    void prepareWSJTXQSO(const QString &receivedCallsign, const QString &grid);
+    void prepareWSJTXQSO(const QString &receivedCallsign, const QString &grid, const QString &id);
     void resetContact();
     void saveContact();

diff --git a/ui/WsjtxWidget.cpp b/ui/WsjtxWidget.cpp
index 024ecd59..9c99c18c 100644
--- a/ui/WsjtxWidget.cpp
+++ b/ui/WsjtxWidget.cpp
@@ -193,7 +193,7 @@ void WsjtxWidget::statusReceived(WsjtxStatus newStatus)
     if ( this->status.dx_call != newStatus.dx_call
          || this->status.dx_grid != newStatus.dx_grid )
     {
-        emit callsignSelected(newStatus.dx_call, newStatus.dx_grid);
+        emit callsignSelected(newStatus.dx_call, newStatus.dx_grid, newStatus.id);
     }

     if ( this->status.mode != newStatus.mode )
diff --git a/ui/WsjtxWidget.h b/ui/WsjtxWidget.h
index 514781b1..1dd35469 100644
--- a/ui/WsjtxWidget.h
+++ b/ui/WsjtxWidget.h
@@ -35,7 +35,7 @@ private slots:
     void actionFilter();

 signals:
-    void callsignSelected(QString callsign, QString grid);
+    void callsignSelected(QString callsign, QString grid, QString id);
     void reply(WsjtxDecode);
     void CQSpot(WsjtxEntry);
     void filteredCQSpot(WsjtxEntry);
```

I can submit a new PR with this change if you agree with the approach and I understood your suggestion correctly.  :)

Thanks!
